### ISSUE TITLE
autograd.backward() segfaults: how to get gradients with respect to a subset of variables in mxnet? 

### ISSUE THREAD

**Issue by altosaar** [apache/incubator-mxnet#8130]
MXNet version: 0.11.0

Python version and distribution: Anaconda 3

## Error Message:
I get erratic segmentation faults from various destructors related to autograd:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xffffffffffffffc0)
  * frame #0: 0x0000000108c6ddeb libmxnet.so`mxnet::exec::FComputeExecutor::~FComputeExecutor() + 59
    frame #1: 0x00007fffaf9a0dae libc++.1.dylib`std::__1::__shared_weak_count::__release_shared() + 44
    frame #2: 0x0000000108c6e456 libmxnet.so`dmlc::any::TypeOnStack<std::__1::vector<std::__1::shared_ptr<mxnet::exec::OpExecutor>, std::__1::allocator<std::__1::shared_ptr<mxnet::exec::OpExecutor> > > >::destroy(dmlc::any::Data*) + 54
    frame #3: 0x0000000108c45db0 libmxnet.so`std::__1::__shared_ptr_emplace<dmlc::any, std::__1::allocator<dmlc::any> >::__on_zero_shared() + 32
    frame #4: 0x00007fffaf9a0dae libc++.1.dylib`std::__1::__shared_weak_count::__release_shared() + 44
    frame #5: 0x0000000108c3f585 libmxnet.so`nnvm::Graph::~Graph() + 69
    frame #6: 0x0000000108c6f71e libmxnet.so`mxnet::exec::GraphExecutor::~GraphExecutor() + 14
    frame #7: 0x0000000108d1d6e8 libmxnet.so`mxnet::autograd::AutogradRuntime::ComputeGradient(std::__1::vector<mxnet::NDArray, std::__1::allocator<mxnet::NDArray> > const&, std::__1::vector<mxnet::NDArray, std::__1::allocator<mxnet::NDArray> > const&, bool, bool) + 7768
    frame #8: 0x0000000108c3ec88 libmxnet.so`MXAutogradBackwardEx + 488
    frame #9: 0x00000001032a82b7 _ctypes.cpython-36m-darwin.so`ffi_call_unix64 + 79
    frame #10: 0x00000001032a8b1b _ctypes.cpython-36m-darwin.so`ffi_call + 875
    frame #11: 0x00000001032a3c70 _ctypes.cpython-36m-darwin.so`_ctypes_callproc + 864
    frame #12: 0x000000010329dda2 _ctypes.cpython-36m-darwin.so`PyCFuncPtr_call + 1106
    frame #13: 0x000000010009ee71 libpython3.6m.dylib`_PyObject_FastCallDict + 257
    frame #14: 0x000000010016da48 libpython3.6m.dylib`call_function + 184
    frame #15: 0x000000010016a528 libpython3.6m.dylib`_PyEval_EvalFrameDefault + 24904
    frame #16: 0x000000010016e98c libpython3.6m.dylib`_PyEval_EvalCodeWithName + 3436
    frame #17: 0x000000010016f311 libpython3.6m.dylib`fast_function + 657
    frame #18: 0x000000010016db21 libpython3.6m.dylib`call_function + 401
    frame #19: 0x000000010016a528 libpython3.6m.dylib`_PyEval_EvalFrameDefault + 24904
    frame #20: 0x000000010016e98c libpython3.6m.dylib`_PyEval_EvalCodeWithName + 3436
    frame #21: 0x000000010016f311 libpython3.6m.dylib`fast_function + 657
    frame #22: 0x000000010016db21 libpython3.6m.dylib`call_function + 401
    frame #23: 0x000000010016a528 libpython3.6m.dylib`_PyEval_EvalFrameDefault + 24904
    frame #24: 0x000000010016e98c libpython3.6m.dylib`_PyEval_EvalCodeWithName + 3436
    frame #25: 0x000000010016f311 libpython3.6m.dylib`fast_function + 657
    frame #26: 0x000000010016db21 libpython3.6m.dylib`call_function + 401
    frame #27: 0x000000010016a5b8 libpython3.6m.dylib`_PyEval_EvalFrameDefault + 25048
    frame #28: 0x000000010016e98c libpython3.6m.dylib`_PyEval_EvalCodeWithName + 3436
    frame #29: 0x000000010016f311 libpython3.6m.dylib`fast_function + 657
    frame #30: 0x000000010016db21 libpython3.6m.dylib`call_function + 401
    frame #31: 0x000000010016a528 libpython3.6m.dylib`_PyEval_EvalFrameDefault + 24904
    frame #32: 0x000000010016e98c libpython3.6m.dylib`_PyEval_EvalCodeWithName + 3436
    frame #33: 0x0000000100164324 libpython3.6m.dylib`PyEval_EvalCode + 100
    frame #34: 0x0000000100199b11 libpython3.6m.dylib`PyRun_FileExFlags + 209
    frame #35: 0x00000001001992ae libpython3.6m.dylib`PyRun_SimpleFileExFlags + 846
    frame #36: 0x00000001001b2066 libpython3.6m.dylib`Py_Main + 3542
    frame #37: 0x0000000100001df8 python`main + 232
    frame #38: 0x00007fffb0dd5235 libdyld.dylib`start + 1
```

## Minimum reproducible example
I'm trying to do variational expectation-maximization in a probabilistic model. 

I need to compute the gradients of terms in the loss with respect to only some parameters.

```
with autograd.record():
  loss = log_q - log joint
  # compute gradients for parameters used to compute log_joint
  (-log_joint).backward(retain_graph=True) 
  # compute score function gradients
  # this leaves gradients of parameters used to compute log_joint unchanged
  neg_log_q.backward()
  # process gradients
trainer.step()
```
To compute `log_joint`, some parameters for computing `log_q` are used, but they are wrapped using `nd.stop_gradient`.

In tensorflow, I would do this:
```
loss = log_q - log_joint
# compute gradients with respect to parameters in log_joint
tf.gradients(-log_joint, log_joint_variables)
# compute gradients with respect to parameters used for log_q
tf.gradients(log_q, log_q_variables)
```

Any help on what I should try to debug this? 

Is there a simpler way to get gradients with respect to only a subset of variables?

**Comment by piiswrong** [comment]
Please post a minimal reproducible example. The code you pasted doesn't run.

No matter what you do, segfault always means a bug.

However, in your case retain_graph isn't the right thing to use.
You can do autograd.backward([-log_joint, neg_log_q])

**Comment by altosaar** [comment]
Thanks @piiswrong !

I installed the newest mxnet (`pip install --pre mxnet`).

Here is a reproducible example based on what you suggested:

```
In [49]: from mxnet import nd, autograd

In [50]: x = nd.array([1.])

In [51]: z = nd.array([1.])

In [52]: x.attach_grad()

In [53]: z.attach_grad()

In [54]: with autograd.record():
    ...:     first = nd.square(x)
    ...:     second = nd.square(z)
    ...:     y = first + second
    ...:     autograd.grad(y, [x], retain_graph=True)
    ...:     autograd.grad(y, [z])
    ...:

In [56]: x.grad
Out[56]:

[ 0.]
<NDArray 1 @cpu(0)>

In [57]: z.grad
Out[57]:

[ 0.]
<NDArray 1 @cpu(0)>
```

**Comment by ZiyueHuang** [comment]
Instead of using two `autograd.grad`, please try
```
print autograd.grad(y, [x, z])
```

**Comment by altosaar** [comment]
Thanks @ZiyueHuang -- that won't work for my use case. Maybe this is a better example:

```
In [49]: from mxnet import nd, autograd

In [50]: x = nd.array([1.])

In [51]: z = nd.array([1.])

In [52]: x.attach_grad()

In [53]: z.attach_grad()

In [64]: with autograd.record():
    ...:     first = nd.square(x)
    ...:     second = nd.square(z)
    ...:     y = first + second
    ...:     autograd.grad(y, [x], retain_graph=True)
    ...:     autograd.grad(second, [z])
    ...:

In [65]: x.grad
Out[65]:

[ 0.]
<NDArray 1 @cpu(0)>

In [66]: z.grad
Out[66]:

[ 0.]
<NDArray 1 @cpu(0)>
```

(I need to differentiate one node for `x` and a different node part of the same computational graph for the gradient of `z`)
